### PR TITLE
#44579 Adding retention option and purging old full and incremental d…

### PR DESCRIPTION
**Pull request by bogdan3d**
…ata that is older than retention days from manifest.

### CODE REVIEW

File: cassandra_backup_service.py [hunk 1]
```diff
@@ -2183,7 +2215,9 @@ def restore(self, columnfamily, nodes, restore_time=None, restore_dir=None, host
                                      help='Optionally provide username to use when connecting to CQLSH')
             repo_parser.add_argument('--cqlsh-pass', dest='cqlsh_pass', required=False,
                                      help='Optionally provide password to use when connecting to CQLSH')
-
+            # Manifest options
+            repo_parser.add_argument('--retention-days', dest='retention_days', required=False, default=30,
```
Comment (houseoftequila) [inline cassandra_backup_service.py:2219]: Can we remove the 30 day default here and by default retain indefinitely.

File: cassandra_backup_service.py [hunk 2]
```diff
@@ -1320,6 +1326,32 @@ def save_manifest(self, keyspace, columnfamily, manifest):
 
         return manifest_file_path
 
+    def manifest_data_retention_pruning(self, data):
```
Comment (houseoftequila) [inline cassandra_backup_service.py:1329]: What is the format of data? Isn't this something like:  `data['incremental'][filename]['created'] = time` per line 1402?.. If so how  does `for inc_ts in data['incremental']:` ... `int(inc_ts)` work?

File: cassandra_backup_service.py [hunk 3]
```diff
@@ -1320,6 +1326,32 @@ def save_manifest(self, keyspace, columnfamily, manifest):
...
+
+        if 'incremental' in data:
+            inc_to_delete = []
+            for inc_ts in data['incremental']:
+                inc_days_old = int(time.mktime(time.gmtime()) - time.mktime(time.gmtime(int(inc_ts)/1000)))/86400
+                if inc_days_old > self.retention_days:
+                    inc_days_old.append(inc_ts)
+
+            for inc_ts in inc_to_delete:
+                del data['incremental'][inc_ts]
```
Comment (houseoftequila) [inline cassandra_backup_service.py:1353]: Lets add a debug line every time we throw out a file with why we are throwing it out and how old it was.

File: cassandra_backup_service.py [hunk 4]
```diff
@@ -1320,6 +1326,32 @@ def save_manifest(self, keyspace, columnfamily, manifest):
...
+        """
+        if 'full' in data:
+            full_to_delete = []
+            for full_ts in data['full']:
+                full_days_old = int(time.mktime(time.gmtime()) - time.mktime(time.gmtime(int(full_ts)/1000)))/86400
+                if full_days_old > self.retention_days:
+                    full_to_delete.append(full_ts)
+
+            for full_ts in full_to_delete:
+                del data['full'][full_ts]
```
Comment (houseoftequila) [inline cassandra_backup_service.py:1343]: Lets add a debug line every time we throw out a file with why we are throwing it out and how old it was.

**Review by houseoftequila — CHANGES_REQUESTED** [review]

File: cassandra_backup_service.py (hunk 2 above)
Comment (bogdan3d) [reply cassandra_backup_service.py:1329]: The data here looks like the following:

```python

    def manifest_data_retention_pruning(self, data):
        """
        Prune provided data to only values within the defined self.retention_days time range.

        :param dict data: manifest data.
        """
        print data
        if 'full' in data:
```

```
{
    u'column_family': u'compaction_history',
    u'keyspace': u'system',
    u'updated': u'2021-01-05 00:58:59',
    u'full': {
        u'1609808336967': {
            u'lb-108-big-Filter.db': {u'created': u'2021-01-05 00:58:58'},
            u'lb-108-big-Summary.db': {u'created': u'2021-01-05 00:58:58'},
            u'lb-108-big-Digest.adler32': {u'created': u'2021-01-05 00:58:58'},
            u'lb-107-big-Index.db': {u'created': u'2021-01-04 23:05:42'},
[...]
```

The `int` casting works on the key which is the timestamp with microseconds.

File: cassandra_backup_service.py (hunk 1 above)
Comment (bogdan3d) [reply cassandra_backup_service.py]: Removed default: https://github.com/onzra/cassandra_backup_service/pull/43/commits/cd0d5b7cfc6ab0f574ad3e781e83734f96719941

File: cassandra_backup_service.py (hunk 2 above)
Comment (houseoftequila) [reply cassandra_backup_service.py:1329]: @bogdan3d what does the incremental data look like?

**Review by bogdan3d — COMMENTED** [review]
Updated incremental logic in this commit https://github.com/onzra/cassandra_backup_service/pull/43/commits/dee5e1d6cdcd3362991ea2be77fa5a6c268276a8 due to different structure. The structure of incremental looks like this in the manifest:

```
data = {
    u'created': u'2021-01-05 00:58:59',
    u'keyspace': u'system',
    'incremental': {
        'lb-37-big-Data.db': {'created': '2021-01-11 22:14:55'},
        'lb-37-big-CompressionInfo.db': {'created': '2021-01-11 22:14:55'},
        'lb-37-big-Statistics.db': {'created': '2021-01-11 22:14:55'},
```

File: cassandra_backup_service.py [hunk 5]
```diff
@@ -1342,6 +1348,37 @@ def save_manifest(self, keyspace, columnfamily, manifest):
...
+                logging.debug('Removing full record from manifest: {0}'.format(full_ts))
+
+        if 'incremental' in data:
+            inc_to_delete = []
+            for inc_fn in data['incremental']:
+                inc_ts = from_human_readable_time(data['incremental'][inc_fn]['created'])
+                print inc_ts
```
Comment (dmertl) [inline cassandra_backup_service.py:1372]: some print statements left in

File: cassandra_backup_service.py [hunk 6]
```diff
@@ -1090,16 +1090,19 @@ class ManifestManager(object):
...
 
-    def __init__(self, cassandra, meta_path, backup_repo):
+    def __init__(self, cassandra, meta_path, backup_repo, retention_days):
         """
         Init.
 
         :param Cassandra cassandra: Cassandra information resource.
+        :param str meta_path: meta path.
         :param BaseBackupRepo backup_repo: backup repository class.
+        :param int retention_days: number of days to retain manifest data.
```
Comment (dmertl) [inline cassandra_backup_service.py:1100]: Should be int or `None` right? As it's optional?

**Pull request merged**